lib/schema/array.js

```diff
@@ -503,30 +491,21 @@ SchemaArray.prototype.clone = function() {
 };
 
 SchemaArray.prototype._castForQuery = function(val, context) {
-  let Constructor = this.casterConstructor;
+  let embeddedSchemaType = this.embeddedSchemaType;
 
   if (val &&
-      Constructor.discriminators &&
-      Constructor.schema &&
-      Constructor.schema.options &&
-      Constructor.schema.options.discriminatorKey) {
-    if (typeof val[Constructor.schema.options.discriminatorKey] === 'string' &&
-        Constructor.discriminators[val[Constructor.schema.options.discriminatorKey]]) {
-      Constructor = Constructor.discriminators[val[Constructor.schema.options.discriminatorKey]];
+      embeddedSchemaType?.discriminators &&
+      typeof embeddedSchemaType?.schema?.options?.discriminatorKey === 'string') {
+    if (embeddedSchemaType.discriminators[val[embeddedSchemaType.schema.options.discriminatorKey]]) {
+      embeddedSchemaType = embeddedSchemaType.discriminators[val[embeddedSchemaType.schema.options.discriminatorKey]];
```

Comment:
Wouldnt it be cleaner to alias this access in a variable?

Reply:
Fixed in https://github.com/Automattic/mongoose/pull/15513/commits/7367171d3c74ff3fd4c77b6016df4ec4f04d74c3